src/core/inductive.ml

```diff
@@ -0,0 +1,195 @@
+open Syntax
+
+(** Logging function for generating of inductive principle. *)
+let log_induc = new_logger 'g' "indu" "generating induction principle"
+let log_induc = log_induc.logger
+
+(** Logging function for rules about inductive type. *)
+let log_induc_rule = new_logger 'l' "irul" "inductive rule"
+let log_induc_rule = log_induc_rule.logger
+
+(** Builtin configuration for induction. *)
+type config =
+  { symb_Prop : sym (** : TYPE. Type of propositions. *)
+  ; symb_prf  : sym (** : Prop → TYPE.
+                        Interpretation of propositions as types. *) }
+
+(** [get_config ss pos] build the configuration using [ss]. *)
+let get_config : Sig_state.t -> Pos.popt -> config = fun ss pos ->
+  let builtin = Builtin.get ss pos in
+  { symb_Prop = builtin "Prop"
+  ; symb_prf  = builtin "P" }
+
+let principle : Sig_state.t -> popt -> sym -> sym list -> term =
+  fun ss pos sind scons_list ->
+  let c = get_config ss pos in
+  let ind = _Symb sind in
+  let prop = _Symb c.symb_Prop in
+  let p = Bindlib.new_var mkfree "p" in
+  let prf_of_p t = _Appl (_Symb c.symb_prf) (_Appl (_Vari p) t) in
+  let app = List.fold_left _Appl in
+
+  let case_of : sym -> tbox = fun scons ->
+    let rec case : tbox list -> term-> tbox = fun xs a ->
+      match unfold a with
+      | Symb(s) ->
+         if s == sind then prf_of_p (app (_Symb scons) (List.rev xs))
+         else fatal pos "%a is not a constructor of %a"
+                pp_symbol scons pp_symbol sind
+      | Prod(a,b) ->
+          let (x,b) =
+            if Bindlib.binder_occur b then
+              Bindlib.unbind b
+            else
+              let x = Bindlib.new_var mkfree "x" in
+              (x, Bindlib.subst b (Vari x))
+          in
+          let b = case ((Bindlib.box_var x)::xs) b in
+          begin
+            match unfold a with
+            | Symb(s) ->
+                let b =
+                  if s == sind then _Impl (prf_of_p (Bindlib.box_var x)) b
+                  else b
+                in
+              _Prod (Bindlib.box a) (Bindlib.bind_var x b)
+            | _ -> fatal pos "The type of %a is not supported"
+                     pp_symbol scons
+          end
+      | _ ->
+          fatal pos "The type of %a is not supported"
+            pp_symbol scons
+    in
+    case [] !(scons.sym_type)
+  in
+
+  let t =
+    let x = Bindlib.new_var mkfree "x" in
+    _Prod ind (Bindlib.bind_var x (prf_of_p (_Vari x)))
+  in
+  (* let add_case t scons = _Impl (case_of scons) t in
+     let t = List.fold_left add_case t scons_list   in *)
+  let rec add_case l r = match l with
+    | []   -> r
+    | t::q -> let t = case_of t in
+              _Impl t (add_case q r)
+  in
+  let t = add_case scons_list t in
+  let t = _Prod (_Impl ind prop) (Bindlib.bind_var p t) in
+  Bindlib.unbox t
+
+let ind_rule : Sig_state.t -> popt -> string -> string -> term -> sym list
+               -> p_rule list =
+  fun _ pos type_name ind_prop_name ind_prop_type cons_list ->
```

Comment:
If you do not use the sig_state, just remove it.
You could factorise the arguments by replacing ind_prop_name and ind_prop_type by the ind_prop symbol.

Reply:
I think it's not possible because the symbol of the induction principle is not yet in the signature at this time. Indeed, here the code tries to build the rules of this induction principle.

Reply:
No, you can first add the symbol in the signature and later add rules.